lib/Schedule/LongSteps/Storage/DynamoDB.pm

```diff
@@ -0,0 +1,538 @@
+has 'id' =>      ( is => 'ro', isa => 'Str', required => 1 );
+
+has 'process_class' => ( is => 'rw', isa => 'Str', required => 1); # rw only for test. Should not changed ever.
+has 'status' =>        ( is => 'rw', isa => 'Str', default => 'pending' );
+has 'what' =>          ( is => 'rw' ,  isa => 'Str', required => 1);
+has 'run_at' =>        ( is => 'rw', isa => 'Maybe[DateTime]', default => sub{ undef; } );
+has 'run_id' =>        ( is => 'rw', isa => 'Maybe[Str]', default => sub{ undef; } );
+has 'state' =>         ( is => 'rw', default => sub{ {}; });
+has 'error' =>         ( is => 'rw', isa => 'Maybe[Str]', default => sub{ undef; } );
+
+# Local to Dynamo Items.
+my $MEMORY_TO_DYNAMO = {
+    id => [ 'id' ],
+    process_class => [ 'process_class' ],
+    status => [ 'mstatus' ], # My status
+    what => [ 'what' ],
+    run_at => [ 'run_at', 'run_at_day' ],
+    run_id => [ 'run_id' ],
+    state => [ 'mstate' ],  # My state
+    error => [ 'merror' ],  # My error
+};
+
+
+=head2 update
+
+Updates via upsert.
+
+=cut
+
+sub update{
+    my ($self, $attributes) = @_;
+
+    foreach my $key ( keys %$attributes ){
+        $self->$key( $attributes->{$key} );
+    }
+    my $dynamo_item = $self->_to_dynamo_item();
+    my $to_update = {};
+    # Only keep attributes that were updated.
+    foreach my $attr ( keys %$attributes ){
+        foreach my $dynamo_attr ( @{ $MEMORY_TO_DYNAMO->{$attr} } ){
+            $to_update->{$dynamo_attr} = $dynamo_item->{$dynamo_attr};
+        }
+    }
+
+    $log->info("Updating Item ID = ".$self->id()." with ".Dumper( $to_update ) );
+
+    my @update_keys = keys %$to_update;
+    my $expression_attributes = {
+        map { ':'.$_ => $to_update->{$_} } @update_keys
+    };
+    # $log->debug("ExpressionAttributes = ".Dumper( $expression_attributes ) );
+    my $update_expression = 'SET '.join(', ', map { $_.' = :'.$_  } @update_keys );
+    # $log->debug("UpdateExpression = ".$update_expression);
+
+    $self->storage()->dynamo_db()->UpdateItem(
+        TableName => $self->storage()->table_name(),
+        Key => { id => { S => $self->id() } },
+        ExpressionAttributeValues => $expression_attributes,
+        UpdateExpression => $update_expression
+    );
+
+    return $self;
+}
+
+=head2 discard_changes
+
+Updates what is updatable.
+
+=cut
+
+sub discard_changes{
+    my ($self) = @_;
+    my $fresh_one = $self->storage()->find_process( $self->id() );
+    foreach my $rw_attr ( qw/process_class status what run_at run_id state error/ ){
+        $self->$rw_attr( $fresh_one->$rw_attr() );
+    }
+    return $self;
+}
+
+sub _insert{
+    my ($self) = @_;
+    my $table_name = $self->storage()->table_name();
+
+    $log->info("Inserting ".ref($self)." , id = ".$self->id()." in Dynamo table ".$table_name);
+    $self->storage()->dynamo_db->PutItem(
+        TableName => $table_name,
+        Item => $self->_to_dynamo_item(),
+    );
+    return $self;
+}
+
+sub _state_encode{
```

Comment:
_state_encode _decode_state, might be worth, fixing the convention (and maybe moving them closer together